mvc/src/main/java/nextstep/mvc/controller/tobe/AnnotationHandlerMapping.java

```diff
@@ -22,9 +31,50 @@ public AnnotationHandlerMapping(Object... basePackage) {
 
     public void initialize() {
         log.info("Initialized AnnotationHandlerMapping!");
+
+        getExecutionMethods()
+            .forEach(this::addHandlerExecution);
+    }
+
+    private void addHandlerExecution(Method method) {
+        RequestMapping annotation = method.getDeclaredAnnotation(RequestMapping.class);
+        String path = annotation.value();
+
+        for (RequestMethod requestMethod : annotation.method()) {
+            HandlerKey handlerKey = new HandlerKey(path, requestMethod);
+            handlerExecutions.put(handlerKey, new HandlerExecution(method));
+
+            log.info("Handler : {} {}", requestMethod.name(), path);
+        }
+    }
+
+    private List<Method> getExecutionMethods() {
+        List<Method> executionMethods = new ArrayList<>();
+
+        for (Object packageName : basePackage) {
+            Reflections reflections = new Reflections(packageName);
+            List<Method[]> declaredMethods = reflections.getTypesAnnotatedWith(Controller.class).stream()
+                .map(Class::getDeclaredMethods)
+                .collect(Collectors.toList());
+
+            addAnnotatedMethods(executionMethods, declaredMethods);
+        }
+        return executionMethods;
+    }
+
+    private void addAnnotatedMethods(List<Method> executionMethods, List<Method[]> foundMethods) {
+        for (Method[] methods : foundMethods) {
+            Arrays.stream(methods)
+                .filter(it -> it.isAnnotationPresent(RequestMapping.class))
+                .forEach(executionMethods::add);
+        }
     }
```

Comment:
`flatMap`을 활용하면 추가 인덴트나 메소드 분리 없이 스트림으로 슉슉 처리할 수 있을 것 같군요😎. 비록 `Controller` 어노테이션을 가져온 후 루프를 돌며 객체를 생성해야 할 것 같기에, 코드가 대거 바뀌긴 할 것 같지만요!

```suggestion
    private List<Method> getExecutionMethods() {
        List<Method> executionMethods = new ArrayList<>();

        for (Object packageName : basePackage) {
            Reflections reflections = new Reflections(packageName);

            reflections.getTypesAnnotatedWith(Controller.class).stream()
                    .map(Class::getDeclaredMethods)
                    .flatMap(Stream::of)
                    .filter(method -> method.isAnnotationPresent(RequestMapping.class))
                    .forEach(executionMethods::add);
        }
        return executionMethods;
    }
```

Reply:
문득 더 줄일 수 있겠다 싶어서 해봤는데... 이건 오히려 너무 안 읽히는 것 같죠? 😅
``` java
return Arrays.stream(basePackage)
                .map(Reflections::new)
                .map(reflections -> reflections.getTypesAnnotatedWith(Controller.class))
                .flatMap(Collection::stream)
                .map(Class::getDeclaredMethods)
                .flatMap(Stream::of)
                .filter(method -> method.isAnnotationPresent(RequestMapping.class))
                .collect(Collectors.toList());
```

Reply:
flatMap에 대해서 잘 알지 못했는데 덕분에 잘 알고 가는 계기가 되었어요.
적용하고 나니 훨씬 깔끔한 코드가 나오는군요! 👍👍👍